File: scheduler.py
#!/usr/bin/env python3
"""
Employee Shift Scheduler - Python version

Features:
- Days: Monday..Sunday
- Shifts: morning, afternoon, evening
- Collects employee preferences (single or ranked)
- Ensures no employee >1 shift/day and <=5 days/week
- Ensures at least 2 employees per shift/day; if fewer, randomly assigns extra (respecting max days)
- Resolves conflicts when preferred shift is full by trying ranked prefs -> other shifts same day -> next day
- Configurable MAX_PER_SHIFT (default 4)
"""

import random
from collections import defaultdict, Counter
import copy

random.seed(42)  # deterministic demo behavior; remove for true randomness

DAYS = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
SHIFTS = ["morning", "afternoon", "evening"]
MAX_PER_SHIFT = 4  # maximum employees allowed per shift per day (defines "full")
MIN_PER_SHIFT = 2  # minimum required per shift per day

# Example demo employees with ranked preferences for each day.
# Format: { "akash": [ ["morning"], ["morning","afternoon"], ... 7 items ] }
# Each day's entry can be a list (ranked) or a single-string (treated as single preferred)
DEMO_EMPLOYEES = {
    "akash":   [["morning"], ["morning"], ["afternoon"], ["morning"], ["evening"], ["morning"], ["morning"]],
    "bipin":   [["morning"], ["morning","afternoon"], ["morning"], ["afternoon"], ["afternoon"], ["evening"], ["afternoon"]],
    "chandu":  [["evening"], ["evening"], ["evening"], ["evening"], ["morning"], ["afternoon"], ["evening"]],
    "damu":    [["afternoon"], ["afternoon"], ["afternoon"], ["afternoon"], ["afternoon"], ["afternoon"], ["afternoon"]],
    "evan":    [["morning","afternoon","evening"]] * 7,
    "farida":  [["afternoon","morning"]] * 7,
    "ganga":   [["evening","morning"]] * 7,
    "harsha":  [["morning"]] * 7,
    "isha":    [["evening"]] * 7,
    "jashu":   [["afternoon"]] * 7,
}


def normalize_prefs(raw):
    """Ensure each entry is a list of ranked preferences for 7 days."""
    prefs = {}
    for name, days in raw.items():
        if len(days) != 7:
            raise ValueError(f"Employee {name} must have 7 day preferences.")
        norm = []
        for d in days:
            if isinstance(d, str):
                norm.append([d])
            elif isinstance(d, list):
                # filter to valid shifts and keep order
                norm.append([s for s in d if s in SHIFTS])
            else:
                raise ValueError("Invalid preference format")
        prefs[name] = norm
    return prefs

class Scheduler:
    def __init__(self, prefs, max_per_shift=MAX_PER_SHIFT, min_per_shift=MIN_PER_SHIFT):
        self.prefs = prefs  # dict name -> list of 7 lists (ranked)
        self.names = list(prefs.keys())
        self.num_emp = len(self.names)
        self.max_per_shift = max_per_shift
        self.min_per_shift = min_per_shift

        # schedule[day_index][shift] = list of employee names
        self.schedule = [ { s: [] for s in SHIFTS } for _ in DAYS ]
        # track days worked per employee
        self.days_worked = Counter()
        # track assignment per day to ensure <=1 shift/day: assigned_today[(day, name)] = True
        self.assigned_today = set()

    def can_assign(self, name, day_idx, shift):
        # not already assigned that day
        if (day_idx, name) in self.assigned_today:
            return False
        # hasn't exceeded 5 days
        if self.days_worked[name] >= 5:
            return False
        # shift capacity
        if len(self.schedule[day_idx][shift]) >= self.max_per_shift:
            return False
        return True

    def assign(self, name, day_idx, shift):
        if not self.can_assign(name, day_idx, shift):
            return False
        self.schedule[day_idx][shift].append(name)
        self.assigned_today.add((day_idx, name))
        self.days_worked[name] += 1
        return True

    def first_pass_preferences(self):
        # Try to assign each employee to their top preference for each day (respect ranked order),
        # but only if not full and employee allowed.
        # We'll iterate employees then days to honor fairness.
        for name in self.names:
            for day_idx in range(7):
                ranked = self.prefs[name][day_idx]
                assigned = False
                for shift in ranked:
                    if self.assign(name, day_idx, shift):
                        assigned = True
                        break
                # if couldn't assign to any preferred shift because full, leave for conflict resolution
        # end

    def enforce_minimums(self):
        # For every day & shift, ensure at least min_per_shift employees. If fewer, randomly pick
        # available employees (not assigned that day and days_worked <5) and assign them.
        for day_idx in range(7):
            for shift in SHIFTS:
                while len(self.schedule[day_idx][shift]) < self.min_per_shift:
                    candidates = [n for n in self.names
                                  if self.can_assign(n, day_idx, shift)]
                    if not candidates:
                        # No one available to fill; break to avoid infinite loop
                        break
                    pick = random.choice(candidates)
                    self.assign(pick, day_idx, shift)

    def resolve_conflicts(self):
        # A more systematic pass: for every employee/day not assigned yet, try to find a slot:
        for name in self.names:
            for day_idx in range(7):
                if (day_idx, name) in self.assigned_today:
                    continue  # already assigned
                # Try ranked prefs for same day
                ranked = self.prefs[name][day_idx]
                assigned = False
                for shift in ranked:
                    if self.can_assign(name, day_idx, shift):
                        self.assign(name, day_idx, shift)
                        assigned = True
                        break
                if assigned:
                    continue
                # try other shifts same day
                for shift in SHIFTS:
                    if shift in ranked: 
                        continue
                    if self.can_assign(name, day_idx, shift):
                        self.assign(name, day_idx, shift)
                        assigned = True
                        break
                if assigned:
                    continue
                # try next day(s)
                for d2 in range(day_idx+1, 7):
                    # try ranked of that next day, then others
                    for shift in self.prefs[name][d2]:
                        if self.can_assign(name, d2, shift):
                            self.assign(name, d2, shift)
                            assigned = True
                            break
                    if assigned:
                        break
                    for shift in SHIFTS:
                        if self.can_assign(name, d2, shift):
                            self.assign(name, d2, shift)
                            assigned = True
                            break
                    if assigned:
                        break
                # If still not assigned, okay — they may have reached 5 days or no slots

    def fill_remaining_minimums(self):
        # After conflict resolution, ensure minimums again (because we may have created new shortages)
        self.enforce_minimums()

    def run(self):
        # Clean state
        self.schedule = [ { s: [] for s in SHIFTS } for _ in DAYS ]
        self.days_worked = Counter()
        self.assigned_today = set()

        # 1) First pass: place people to their preferences as possible
        self.first_pass_preferences()
        # 2) Ensure minimum staffing
        self.enforce_minimums()
        # 3) Resolve conflicts for unassigned employees
        self.resolve_conflicts()
        # 4) Final pass to ensure minimums again
        self.fill_remaining_minimums()

    def pretty_print(self):
        out_lines = []
        header = "Day      | Morning             | Afternoon           | Evening"
        out_lines.append(header)
        out_lines.append("-"*len(header))
        for i, day in enumerate(DAYS):
            row = f"{day:<8}| "
            for shift in SHIFTS:
                names = ", ".join(self.schedule[i][shift]) if self.schedule[i][shift] else "(none)"
                row += f"{names:<20}| "
            out_lines.append(row)
        # Also output days worked
        out_lines.append("\nDays worked per employee (name: days):")
        for name in sorted(self.names):
            out_lines.append(f" - {name}: {self.days_worked[name]}")
        return "\n".join(out_lines)


def demo_run():
    prefs = normalize_prefs(DEMO_EMPLOYEES)
    s = Scheduler(prefs)
    s.run()
    print(s.pretty_print())

if __name__ == "__main__":
    print("=== Employee Scheduler (Python demo) ===")
    print("Using demo employees. To use custom input, modify DEMO_EMPLOYEES in the script.")
    demo_run()

File: scheduler.cpp
// scheduler.cpp
// Employee Shift Scheduler - C++ version
//
// Similar behavior to the Python script.
// Build: g++ -std=c++17 scheduler.cpp -o scheduler
// Run: ./scheduler

#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>
#include <random>
#include <ctime>
#include <set>
#include <iomanip>   // for setw and formatting
using namespace std;


const vector<string> DAYS = {"Mon","Tue","Wed","Thu","Fri","Sat","Sun"};
const vector<string> SHIFTS = {"morning","afternoon","evening"};
int MAX_PER_SHIFT = 4;
int MIN_PER_SHIFT = 2;
std::mt19937 rng(42);

using Prefs = map<string, vector<vector<string>>>; // name -> 7-day list of ranked prefs

Prefs demo_prefs() {
    Prefs p;
    p["akash"]  = { {"morning"},{"morning"},{"afternoon"},{"morning"},{"evening"},{"morning"},{"morning"} };
    p["bipin"]  = { {"morning"},{"morning","afternoon"},{"morning"},{"afternoon"},{"afternoon"},{"evening"},{"afternoon"} };
    p["chandu"] = { {"evening"},{"evening"},{"evening"},{"evening"},{"morning"},{"afternoon"},{"evening"} };
    p["damu"]   = { {"afternoon"},{"afternoon"},{"afternoon"},{"afternoon"},{"afternoon"},{"afternoon"},{"afternoon"} };
    p["evan"]   = vector<vector<string>>(7, {"morning","afternoon","evening"});
    p["farida"] = vector<vector<string>>(7, {"afternoon","morning"});
    p["ganga"]  = vector<vector<string>>(7, {"evening","morning"});
    p["harsha"] = vector<vector<string>>(7, {"morning"});
    p["isha"]   = vector<vector<string>>(7, {"evening"});
    p["jashu"]  = vector<vector<string>>(7, {"afternoon"});
    return p;
}
struct Scheduler {
    Prefs prefs;
    vector<string> names;
    // schedule[day][shift] = vector<string>
    vector<map<string, vector<string>>> schedule;
    unordered_map<string,int> days_worked;
    set<pair<int,string>> assigned_today;

    Scheduler(const Prefs &p) : prefs(p) {
        for (auto &kv : prefs) names.push_back(kv.first);
        schedule.assign(7, map<string, vector<string>>());
        for (int i=0;i<7;i++){
            for (auto &s : SHIFTS) schedule[i][s] = vector<string>();
        }
        for (auto &n : names) days_worked[n] = 0;
    }

    bool can_assign(const string &name, int day, const string &shift) {
        if (assigned_today.count({day,name})) return false;
        if (days_worked[name] >= 5) return false;
        if ((int)schedule[day][shift].size() >= MAX_PER_SHIFT) return false;
        return true;
    }

    bool assign(const string &name, int day, const string &shift) {
        if (!can_assign(name, day, shift)) return false;
        schedule[day][shift].push_back(name);
        assigned_today.insert({day,name});
        days_worked[name]++;
        return true;
    }

    void first_pass_preferences() {
        for (auto &name : names) {
            for (int d=0; d<7; ++d) {
                auto &ranked = prefs.at(name)[d];
                for (auto &shift : ranked) {
                    if (assign(name, d, shift)) break;
                }
            }
        }
    }

    void enforce_minimums() {
        for (int d=0; d<7; ++d) {
            for (auto &shift : SHIFTS) {
                while ((int)schedule[d][shift].size() < MIN_PER_SHIFT) {
                    vector<string> candidates;
                    for (auto &n : names) {
                        if (can_assign(n,d,shift)) candidates.push_back(n);
                    }
                    if (candidates.empty()) break;
                    uniform_int_distribution<int> dist(0, (int)candidates.size()-1);
                    string pick = candidates[dist(rng)];
                    assign(pick, d, shift);
                }
            }
        }
    }

    void resolve_conflicts() {
        for (auto &name : names) {
            for (int d=0; d<7; ++d) {
                if (assigned_today.count({d,name})) continue;
                bool assigned = false;
                auto &ranked = prefs.at(name)[d];
                for (auto &shift : ranked) {
                    if (can_assign(name,d,shift) && assign(name,d,shift)) { assigned = true; break; }
                }
                if (assigned) continue;
                for (auto &shift : SHIFTS) {
                    if (can_assign(name,d,shift) && assign(name,d,shift)) { assigned = true; break; }
                }
                if (assigned) continue;
                for (int d2=d+1; d2<7 && !assigned; ++d2) {
                    for (auto &shift : prefs.at(name)[d2]) {
                        if (can_assign(name,d2,shift) && assign(name,d2,shift)) { assigned = true; break; }
                    }
                    for (auto &shift : SHIFTS) {
                        if (assigned) break;
                        if (can_assign(name,d2,shift) && assign(name,d2,shift)) { assigned = true; break; }
                    }
                }
            }
        }
    }

    void run() {
        // reset
        for (int i=0;i<7;i++){
            for (auto &s : SHIFTS) schedule[i][s].clear();
        }
        for (auto &kv : days_worked) kv.second = 0;
        assigned_today.clear();

        first_pass_preferences();
        enforce_minimums();
        resolve_conflicts();
        enforce_minimums();
    }

    void print_schedule() {
        cout << left << setw(8) << "Day" << "| " << setw(22) << "Morning" << "| " << setw(22) << "Afternoon" << "| " << "Evening" << "\n";
        cout << string(70,'-') << "\n";
        for (int d=0; d<7; ++d) {
            cout << left << setw(8) << DAYS[d] << "| ";
            for (auto &shift : SHIFTS) {
                string joined = "(none)";
                if (!schedule[d][shift].empty()) {
                    string s;
                    for (size_t i=0;i<schedule[d][shift].size();++i){
                        if (i) s += ", ";
                        s += schedule[d][shift][i];
                    }
                    joined = s;
                }
                cout << setw(22) << joined << "| ";
            }
            cout << "\n";
        }
        cout << "\nDays worked per employee:\n";
        for (auto &name : names) {
            cout << " - " << name << ": " << days_worked[name] << "\n";
        }
    }
};

int main(){
    cout << "=== Employee Scheduler (C++ demo) ===\n";
    auto p = demo_prefs();
    Scheduler s(p);
    s.run();
    s.print_schedule();
    return 0;
}
